# Please visit the [wiki page](https://github.com/Tanguay-Lab/Manuscripts/wiki/Dasgupta_2021_Front_Cell_Dev_Biol)

# Phenotypically Anchored mRNA and miRNA Expression Profiling in Zebrafish Reveals Flame Retardant Chemical Toxicity Networks
Subham Dasgupta<sup>1</sup>, Cheryl L. Dunham<sup>1</sup>, Lisa Truong<sup>1</sup>, Michael T. Simonich<sup>1</sup>, Christopher M. Sullivan<sup>2</sup>, and Robyn L. Tanguay<sup>1</sup>
<br><br>
### Author affiliations:
  1. The Sinnhuber Aquatic Research Laboratory, Department of Environmental and Molecular Toxicology, Oregon State University, Corvallis, OR, United States
  2. Center for Genome Research and Computing, Oregon State University, Corvallis, OR, United States

<br>

## Citation
Please cite the following if you use information or files from this repository:
<br><br><br>
Dasgupta S, Dunham CL, Truong L, Simonich MT, Sullivan CM, and Tanguay RL. (2021). Phenotypically Anchored mRNA and miRNA Expression Profiling in Zebrafish Reveals Flame Retardant Chemical Toxicity Networks. *Frontiers in Cell and Developmental Biology*. **9**:663032. 
doi: [10.3389/fcell.2021.663032](https://www.frontiersin.org/articles/10.3389/fcell.2021.663032/)

<br>

Gene Expression Omnibus (GEO) Accession Number: [GSE169013](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE169013)
<br>
Sequence Read Archive (SRA) Accession Number: [SRP310924](https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP310924)
<br>
PubMed ID: [33898466](https://pubmed.ncbi.nlm.nih.gov/33898466/)
<br>
PubMed Central ID: [PMC8063052](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8063052/)

<br>

## Abstract
The ubiquitous use of flame retardant chemicals (FRCs) in the manufacture of many consumer products leads to inevitable environmental releases and human exposures. Studying toxic effects of FRCs as a group is challenging since they widely differ in physicochemical properties. We previously used zebrafish as a model to screen 61 representative FRCs and showed that many induced behavioral and teratogenic effects, with aryl phosphates identified as the most active. In this study, we selected 10 FRCs belonging to diverse physicochemical classes and zebrafish toxicity profiles to identify the gene expression responses following exposures. For each FRC, we executed paired mRNA-micro-RNA (miR) sequencing, which enabled us to study mRNA expression patterns and investigate the role of miRs as posttranscriptional regulators of gene expression. We found widespread disruption of mRNA and miR expression across several FRCs. Neurodevelopment was a key disrupted biological process across multiple FRCs and was corroborated by behavioral deficits. Several mRNAs (e.g., osbpl2a) and miRs (e.g., mir-125b-5p), showed differential expression common to multiple FRCs (10 and 7 respectively). These common miRs were also predicted to regulate a network of differentially expressed genes with diverse functions, including apoptosis, neurodevelopment, lipid regulation and inflammation. Commonly disrupted transcription factors (TFs) such as retinoic acid receptor, retinoid X receptor, and vitamin D regulator were predicted to regulate a wide network of differentially expressed mRNAs across a majority of the FRCs. Many of the differential mRNA-TF and mRNA-miR pairs were predicted to play important roles in development as well as cancer signaling. Specific comparisons between TBBPA and its derivative TBBPA-DBPE showed contrasting gene expression patterns that corroborated with their phenotypic profiles. The newer generation FRCs such as IPP and TCEP produced distinct gene expression changes compared to the legacy FRC BDE-47. Our study is the first to establish a mRNA-miR-TF regulatory network across a large group of structurally diverse FRCs and diverse phenotypic responses. The purpose was to discover common and unique biological targets that will help us understand mechanisms of action for these important chemicals and establish this approach as an important tool for better understanding toxic effects of environmental contaminants.

<br>

## Detailed Methods
[Detailed RNA Sequencing Analysis Methods](https://github.com/Tanguay-Lab/Manuscripts/blob/main/Dasgupta_et._al._(2021)_Front_Cell_Dev_Biol/FRC_Manuscript_Methods_NGS_Analysis.md)

<br>

## Supplemental Files
Supplemental File [FRC_adapters.fa](https://github.com/Tanguay-Lab/Manuscripts/blob/main/Dasgupta_et._al._(2021)_Front_Cell_Dev_Biol/Files/FRC_adapters.fa)

Supplemental File [hisat2_extract_snps_haplotypes_VCF_CD.py](https://github.com/Tanguay-Lab/Manuscripts/blob/main/Dasgupta_et._al._(2021)_Front_Cell_Dev_Biol/Files/hisat2_extract_snps_haplotypes_VCF_CD.py)

Supplemental File [Differential_gene_analysis_custom_script.R](https://github.com/Tanguay-Lab/Manuscripts/blob/main/Dasgupta_et._al._(2021)_Front_Cell_Dev_Biol/Files/Differential_gene_analysis_custom_script.R)

Supplemental File [FRC_miRNA_adapters.fa](https://github.com/Tanguay-Lab/Manuscripts/blob/main/Dasgupta_et._al._(2021)_Front_Cell_Dev_Biol/Files/FRC_miRNA_adapters.fa)

Supplemental File [mapper_CLD_Q33.pl](https://github.com/Tanguay-Lab/Manuscripts/blob/main/Dasgupta_et._al._(2021)_Front_Cell_Dev_Biol/Files/mapper_CLD_Q33.pl)

Supplemental File [FRCs_miRNA_config_file.txt](https://github.com/Tanguay-Lab/Manuscripts/blob/main/Dasgupta_et._al._(2021)_Front_Cell_Dev_Biol/Files/FRCs_miRNA_config_file.txt)

Supplemental File [make_html_2012_CLD.pl](https://github.com/Tanguay-Lab/Manuscripts/blob/main/Dasgupta_et._al._(2021)_Front_Cell_Dev_Biol/Files/make_html_2012_CLD.pl)

Supplemental File [miRDeep2_2012_CLD.pl](https://github.com/Tanguay-Lab/Manuscripts/blob/main/Dasgupta_et._al._(2021)_Front_Cell_Dev_Biol/Files/miRDeep2_2012_CLD.pl)

Supplemental File [Differential_miRNA_analysis_custom_script.R](https://github.com/Tanguay-Lab/Manuscripts/blob/main/Dasgupta_et._al._(2021)_Front_Cell_Dev_Biol/Files/Differential_miRNA_analysis_custom_script.R)
